README.md
# DateTime Locale Patterns

[![Latest version](https://img.shields.io/npm/v/datetime-locale-patterns)
 ![Dependency status](https://img.shields.io/librariesio/release/npm/datetime-locale-patterns)
](https://www.npmjs.com/package/datetime-locale-patterns)
[![Coverage](https://codecov.io/gh/prantlf/datetime-locale-patterns/branch/master/graph/badge.svg)](https://codecov.io/gh/prantlf/datetime-locale-patterns)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/9f1034029c0747a980cd49f64f16338b)](https://www.codacy.com/app/prantlf/datetime-locale-patterns?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=prantlf/datetime-locale-patterns&amp;utm_campaign=Badge_Grade)

Provides localized date/time format patterns for styles `full`, `long`, `medium` and `short`, usable with [`Intl.DateTimeFormat`], compliant with [Unicode LDML]. For example:

| Locale | Style | Pattern             | Example                     |
|:-------|:------|:--------------------|:----------------------------|
| en     | short | M/d/yy, h:mm a      | 2/3/01, 4:05 AM             |
| cs     | long  | d. MMMM y H:mm:ss z | 3. února 1901 4:05:06 GMT+1 |

```js
const formatter = new Intl.DateTimeFormat('cs', { dateStyle: 'short'})
const pattern = getDateTimeLocalePattern(formatter) // d.M.yy
```

* ES, CJS and UMD module exports.
* TypeScript type declarations (typings).
* No other dependencies.
* Tiny code base - 1.32 kB minified, 586 B gzipped, 506 B brotlied. Bundle with all data - 115 kB minified, 7.27 kB gzipped, 5.83 kB brotlied.
* Generated from the official [CLDR data] version 44.0.1. 710 locales supported.
* Resolves four date/time-formatting pattern styles (lengths) - `full`, `long`, `medium`, `short`.

This library uses Unicode data to compute the patterns. It means that it does not need a working implementation of `Intl.DateTimeFormat`, but its size is increased by the size of the data. If you are interested in a library, which uses `Intl.DateTimeFormat` and needs no data, have look at [intl-datetimeformat-options].

Related projects:

* [date-and-time-formatter] - formats a date/time value to a localised string using a pattern consisting of [Unicode LDML] tokens.
* [datetime-placeholder] - simplifies a date/time-formatting pattern using [Unicode LDML] tokens to a pattern usable in date/time pickers.
* [intl-datetimeformat-options] - provides localized date/time format patterns for styles `full`, `long`, `medium` and `short`, using [`Intl.DateTimeFormat`].
* [intl-datetimeformat-pattern] - creates a valid [`Intl.DateTimeFormat`] options object from a [Unicode CLDR] skeleton or token pattern.

## Motivation

When do you need to know the date/time format pattern? When is just formatting a date or time values to a string not enough?

1. **Date/time pickers.** You format with [`Intl.DateTimeFormat`]. [`Intl.DateTimeFormat`] or [luxon] format using a localized pattern decided by a specified locale. No need to provide the format explicitly. But what if you need to display the format pattern, which is used for the formatting? For example, in a date picker field as a value placeholder.
2. **Raw date/time formatting.** You do not format with [`Intl.DateTimeFormat`]. Libraries like [date-and-time], [date-fns] and others format using patterns provided by the developer. But how to get a localized pattern for a particular language and country?

There is no built-in API in browsers or Node.js for getting localized date/time-formatting patterns.

## Synopsis

1. **Date/time pickers.** Get a pattern to see in an abstract way how a concrete date will be formatted:

```js
import { getDateTimeLocalePattern } from 'datetime-locale-patterns'

const date = new Date(1, 1, 3, 4, 5, 6)
const options = { dateStyle: 'short', timeStyle: 'short' }

const formatter = new Intl.DateTimeFormat('en', options)
const text = formatter.format(date)
console.log(text) // prints '2/3/01, 4:05 AM'

const pattern = getDateTimeLocalePattern('en', options)
console.log(pattern) // prints 'M/d/yy, h:mm a'
```

2. **Raw date/time formatting.** Get a pattern to format a concrete date with:

```js
import { getDateTimeLocalePattern } from 'datetime-locale-patterns'
import formatter from 'date-and-time'

const date = new Date(1, 1, 3, 4, 5, 6)

const pattern = getDateTimeLocalePattern('en', 'short', 'short')
console.log(pattern) // prints 'M/d/yy, h:mm a'

const text = formatter.format(date, pattern)
console.log(text) // prints '2/3/01, 4:05 AM'
```

## Installation

This module can be installed in your project using [NPM], [PNPM] or [Yarn]. Make sure, that you use [Node.js] version 16.14 or newer.

```sh
$ npm i datetime-locale-patterns
$ pnpm i datetime-locale-patterns
$ yarn add datetime-locale-patterns
```

Functions are exposed as named exports from ES and CJS modules, for example:

```js
import { getDateTimeLocalePattern } from 'datetime-locale-patterns'
```

```js
const { getDateTimeLocalePattern } = require('datetime-locale-patterns')
```

A UMD module can be loaded to the browser either directly:

```html
<script src="https://unpkg.com/datetime-locale-patterns@1.0.0/lib/index.min.js"></script>
<script>
  const { getDateTimeLocalePattern } = window.dateTimeLocalePatterns
</script>
```

Or using an AMD module loader:

```html
<script>
  require([
    'https://unpkg.com/datetime-locale-patterns@1.0.0/lib/index.min.js'
  ], ({ getDateTimeLocalePattern }) => {
    ...
  })
</script>
```

## API

This library works well with [`Intl.DateTimeFormat`] by using concepts from [Unicode CLDR] (Common Locale Data Repository):

* It expects [Unicode locales] (2-letter and 3-letter languages and countries from [ISO 639] and [ISO 3166], formatted according to [extended BCP 47]), which [`Intl` locales] is based on.
* It supports [Unicode calendar formats] - `date`, `time`, `dateTime`.
* It expects [Unicode pattern styles] (lengths) - `full`, `long`, `medium`, `short`, as `dateStyle` and `timeStyle` properties of [`Intl.DateTimeFomat` options] do, or an instance of `Intl.DateTimeFomat`.
* It supplies date/time-formatting patterns using [date fields] from [Unicode LDML] (Locale Data Markup Language) as tokens.

### getDateLocalePattern(locale, style?)

Returns a pattern to format a date-only value for the specified locale with the specified style.

The **locale** argument can be either a locale or an instance of `Intl.DateTimeFomat`. The second argument won't be needed in the latter case.

The **style** argument can be either a string (`full`, `long`, `medium`, `short`) or an object with the `dateStyle` property from [`Intl.DateTimeFomat` options].

```js
import { getDateLocalePattern } from 'datetime-locale-patterns'

const pattern = getDateLocalePattern('en', 'short')
console.log(pattern) // prints 'M/d/yy'
```

### getTimeLocalePattern(locale, style?)

Returns a pattern to format a time-only value for the specified locale with the specified style.

The **locale** argument can be either a locale or an instance of `Intl.DateTimeFomat`. The second argument won't be needed in the latter case.

The **style** argument can be either a string (`full`, `long`, `medium`, `short`) or an object with the `timeStyle` property from [`Intl.DateTimeFomat` options].

```js
import { getLocalePattern } from 'datetime-locale-patterns'

const pattern = getTimeLocalePattern('en', 'short')
console.log(pattern) // prints 'h:mm a'
```

### getDateTimeLocalePattern(locale, dateStyle?, timeStyle?)

Returns a pattern to format a date+time value for the specified locale with the specified style.

The **locale** argument can be either a locale or an instance of `Intl.DateTimeFomat`. The second and third arguments won't be needed in the latter case.

The **dateStyle** argument can be either a string (`full`, `long`, `medium`, `short`) or an object with `dateStyle` and `timeStyle` properties from [`Intl.DateTimeFomat` options]. If it is a string, the **timeStyle** argument will be required (as a string).

```js
import { getDateTimeLocalePattern } from 'datetime-locale-patterns'

const pattern = getDateTimeLocalePattern('en', 'short', 'short')
console.log(pattern) // prints 'M/d/yy, h:mm a'
```

### setLocalePatterns(data)

Initializes the data used by other functions to date/time-formatting patterns.

The default data including all locales, formats and pattern styles are exposed as the `datetime-locale-patterns/data/all` module. Smaller data including all locales and formats for the short pattern style are exposed as the `datetime-locale-patterns/data/short` module. [Custom data](#custom-patterns) can be generated using the `create-datetime-locale-patterns` script.

```js
import { setLocalePatterns } from 'datetime-locale-patterns/code'
import patterns from 'datetime-locale-patterns/data/short' assert { type: 'json' }

setLocalePatterns(patterns) // use only short date/time-formatting patterns
```

## Custom Patterns

Patterns may need to be customised, if you need them to differ from the CLDR standard, or if you need to support only a subset of locales, formats or styles to reduce the size of the data loaded by your application.

The default data including all locales, formats and pattern styles are exposed as the `datetime-locale-patterns/data/all` module. Smaller data including all locales and formats for the short pattern style are exposed as the `datetime-locale-patterns/data/short` module. Other data (list of localized date/time format patterns) can be generated using the `create-datetime-locale-patterns` script and enabled using the `setLocalePatterns` function.

### Creating Custom Data

Format of the data is JSON:

    {
      formats: [...], // a subset of date, time, dateTime
      styles: [...],  // a subset of full, long, medium, short
      patterns: {
        // locale to array of formats; a format is an array of styles
        [locale]: [[date], [time], [date-time]],
        ...
      }
    }

For example, for supporting only `en` and `cs` locales and `short` and `long` styles, the following command line will create the following file content:

    create-datetime-locale-patterns -l en,cs -s short,long

```json
{
  "formats": ["date", "time", "dateTime"],
  "styles": ["short", "long"],
  "patterns": {
    "en": [["M/d/yy", "MMMM d, y"], ["h:mm a", "h:mm:ss a z"], ["{1}, {0}", "{1}, {0}"]],
    "cs": [["dd.MM.yy", "d. MMMM y"], ["H:mm", "H:mm:ss z"], ["{1} {0}", "{1} {0}"]]
  }
}
```

The command-line script `create-datetime-locale-patterns` is installed to the `bin` directory in `node_modules`:

    Usage: create-datetime-locale-patterns [options]

    Options:
      -l|--locales <>  list of locales to include (default: all)
      -f|--formats <>  list of format patterns (default: date,time,dateTime)
      -s|--styles <>   list of format styles (default: short,medium,long,full)
      -o|--output <>   output file (default: console)
      -p|--pretty      prettify the JSON output (default: minified)
      -V|--version     print version number
      -h|--help        print usage instructions

    Examples:
      $ create-datetime-locale-patterns -s short,long -o patterns.json
      $ create-datetime-locale-patterns -l en,en-GB,cs,de,de-AT -f date -p

### Loading Custom Data

If you want to use the limited data from the `datetime-locale-patterns/data/short` module or your custom data, do not import functions from the default `datetime-locale-patterns` module, but from the `datetime-locale-patterns/code` module, which will not load the default data automatically. Once you do it, you will have to supply the custom data to the `setLocalePatterns` function before you call any other function from this library:

```js
import { setLocalePatterns, getDateTimeLocalePattern } from 'datetime-locale-patterns/code'
import patterns from './patterns.json' assert { type: 'json' }

setLocalePatterns(patterns)
const pattern = getDateTimeLocalePattern('cs', 'short', 'short')
```

## Problems

The assumption that [`Intl.DateTimeFormat`] uses patterns from [Unicode CLDR] may not be always true.

* If the date/time formatting provided by the underlying OS API is used, the actual pattern may differ, although the multi-platform [Unicode CLDR] is supposed to include the same localized patterns.
* If [ICU] (International Components for Unicode) library, which is supposed to be in sync with [Unicode CLDR], is used by a browser or Node.js, the version of that [ICU] may differ from the latest version of [Unicode CLDR]. This can be seen in `long` and `full` styles of the `en` locale, which include a different "glue pattern" for joining data and time patterns. [ICU] includes " at ", while [Unicode CLDR] includes ",".

JavaScript platforms using [ICU] will match the patterns returned by this library well. It means Node.js 13 and newer and browsers Chromium and Firefox. Safari doesn't use [ICU], AFAIK.

## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding style.  Add unit tests for any new or changed functionality. Lint and test your code using Grunt.

## License

Copyright (c) 2023 Ferdinand Prantl

Licensed under the MIT license.

[Node.js]: http://nodejs.org/
[NPM]: https://www.npmjs.com/
[PNPM]: https://pnpm.io/
[Yarn]: https://yarnpkg.com/
[date-and-time]: https://github.com/knowledgecode/date-and-time
[date-fns]: https://github.com/date-fns/date-fns
[luxon]: https://moment.github.io/luxon/
[ICU]: https://icu.unicode.org/
[CLDR data]: https://www.npmjs.com/package/cldr-dates-full
[Unicode LDML]: https://unicode.org/reports/tr35/
[Unicode CLDR]: https://cldr.unicode.org/
[Unicode locales]: https://www.unicode.org/reports/tr35/#Language_and_Locale_IDs
[`Intl` locales]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#locale_identification_and_negotiation
[ISO 3166]: https://en.wikipedia.org/wiki/ISO_3166
[ISO 639]: https://en.wikipedia.org/wiki/ISO_639
[extended BCP 47]: https://cldr.unicode.org/index/bcp47-extension
[`Intl.DateTimeFormat`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat
[Unicode pattern styles]: https://unicode.org/reports/tr35/tr35-dates.html#24-element-dateformats
[Unicode calendar formats]: https://unicode.org/reports/tr35/tr35-dates.html#2-calendar-elements
[date fields]: http://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
[`Intl.DateTimeFomat` options]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat#options
[date-and-time-formatter]: https://github.com/prantlf/date-and-time-formatter
[datetime-placeholder]: https://github.com/prantlf/datetime-placeholder
[intl-datetimeformat-pattern]: https://github.com/caridy/intl-datetimeformat-pattern
[intl-datetimeformat-options]: https://github.com/prantlf/intl-datetimeformat-options


CHANGELOG.md
## [1.0.2](https://github.com/prantlf/datetime-locale-patterns/compare/v1.0.1...v1.0.2) (2023-12-10)


### Bug Fixes

* Update dependencies ([f6a9592](https://github.com/prantlf/datetime-locale-patterns/commit/f6a95921cea0714491f866a13f296f7c5ca04fe7))

## [1.0.1](https://github.com/prantlf/datetime-locale-patterns/compare/v1.0.0...v1.0.1) (2023-03-16)


### Bug Fixes

* Simplify TypeScript types ([25f5351](https://github.com/prantlf/datetime-locale-patterns/commit/25f5351f0dad53980d582e7bec8929315f8d96e5))

# 1.0.0 (2023-01-02)
 
Initial release


test/types.test.ts
import {
  getDateLocalePattern, getTimeLocalePattern, getDateTimeLocalePattern, PatternStyle
} from '../lib/index.js'
import { setLocalePatterns } from '../lib/code.js'

declare type testCallback = () => void
declare function test (label: string, callback: testCallback): void

test('Type declarations for TypeScript', () => {
  const patternStyle: PatternStyle = 'short'
  let _pattern: string
  _pattern = getDateLocalePattern('cs', patternStyle)
  _pattern = getDateLocalePattern('cs', { dateStyle: patternStyle })
  _pattern = getTimeLocalePattern('cs', patternStyle)
  _pattern = getTimeLocalePattern('cs', { timeStyle: patternStyle })
  _pattern = getDateTimeLocalePattern('cs', patternStyle, patternStyle)
  _pattern = getDateTimeLocalePattern('cs', { dateStyle: patternStyle, timeStyle: patternStyle })
  setLocalePatterns({ formats: [], styles: [], patterns: {} })
  const formatter = new Intl.DateTimeFormat('cs')
  _pattern = getDateLocalePattern(formatter)
  _pattern = getTimeLocalePattern(formatter)
  _pattern = getDateTimeLocalePattern(formatter)
})


test/all-data.test.js
import tehanu from 'tehanu'
import { deepStrictEqual, fail } from 'assert'
import data from '../lib/all.json' assert { type: 'json' }
import availableLocales from 'cldr-core/availableLocales.json' assert { type: 'json' }

const test = tehanu(import.meta.url)
const locales = availableLocales.availableLocales.full.map(locale => locale.toLowerCase())
const { formats, styles, patterns } = data

test('provides patterns for all formats', () => {
  deepStrictEqual(formats, ['date', 'time', 'dateTime'])
})

test('provides patterns for all styles', () => {
  deepStrictEqual(styles, ['full', 'long', 'medium', 'short'])
})

test('provides patterns for all locales', () => {
  for (const locale of locales) {
    if (!patterns[locale]) fail(`missing locale "${locale}"`)
  }
})


test/set-data.test.js
import tehanu from 'tehanu'
import { strictEqual, throws } from 'assert'
import { getDateTimeLocalePattern, setLocalePatterns } from 'datetime-locale-patterns/code'
import { format } from 'date-fns'
import shortPatterns from 'datetime-locale-patterns/data/short' assert { type: 'json' }
import allPatterns from 'datetime-locale-patterns/data/all' assert { type: 'json' }

const test = tehanu(import.meta.url)
const date = new Date(1, 1, 3, 4, 5, 6) // 1901-02-03 04:05:06

test.before(() => setLocalePatterns(shortPatterns))
test.after(() => setLocalePatterns(allPatterns))

test('supports short format pattern', () => {
  const pattern = getDateTimeLocalePattern('cs', 'short', 'short')
  strictEqual(pattern, 'dd.MM.yy H:mm')
  strictEqual(format(date, pattern), '03.02.01 4:05')
})

test('does not support long format pattern', () => {
  throws(() => getDateTimeLocalePattern('cs', 'long', 'long'))
})


lib/code.js
let formatIndexes, styleIndexes, patternsByLocale

// Convert an array of names to an object, where the names are used as keys
// pointing to values representing indexes of the names in the original array
function initializeIndexes(names) {
  const indexes = {}
  for (let i = 0, l = names.length; i < l; ++i) indexes[names[i]] = i
  return indexes
}

export function setLocalePatterns(data) {
  const { formats, styles, patterns } = data

  formatIndexes = initializeIndexes(formats)
  styleIndexes = initializeIndexes(styles)

  patternsByLocale = patterns
}

// Enforce lower-case and dash as a delimiter
function normalizeLocale(locale) {
  return locale.toLowerCase().replace('_', '-')
}

// Cut the first part of the locale before the first dash (language)
function getLanguage(locale) {
  const separator = locale.indexOf('-')
  return separator > 0 && locale.substring(0, separator)
}

function getPattern(locale, style, format) {
  locale = normalizeLocale(locale)

  // If language with country is not supported, try the language alone
  let patterns = patternsByLocale[locale]
  if (!patterns) {
    const language = getLanguage(locale)
    if (language) patterns = patternsByLocale[language]
    if (!patterns) throw new Error(`Invalid locale: "${locale}".`)
  }

  const formatIndex = formatIndexes[format]
  if (formatIndex === undefined) throw new Error(`Invalid format: "${format}".`)

  const styleIndex = styleIndexes[style]
  if (styleIndex === undefined) throw new Error(`Invalid style: "${style}".`)

  return patterns[formatIndex][styleIndex]
}

export function getDateLocalePattern(locale, style) {
  // Get locale and dateStyle from any argument combination
  if (typeof locale !== 'string') {
    ({ locale, dateStyle: style } = locale.resolvedOptions())
  } else if (typeof style !== 'string') {
    style = style.dateStyle
  }

  return getPattern(locale, style, 'date')
}

export function getTimeLocalePattern(locale, style) {
  // Get locale and timeStyle from any argument combination
  if (typeof locale !== 'string') {
    ({ locale, timeStyle: style } = locale.resolvedOptions())
  } else if (typeof style !== 'string') {
    style = style.timeStyle
  }

  return getPattern(locale, style, 'time')
}

export function getDateTimeLocalePattern(locale, dateStyle, timeStyle) {
  // Get locale, dateStyle and timeStyle from any argument combination
  if (typeof locale !== 'string') {
    ({ locale, dateStyle, timeStyle } = locale.resolvedOptions())
  } else if (typeof dateStyle !== 'string') {
    ({ dateStyle, timeStyle } = dateStyle)
  }

  const datePattern = getDateLocalePattern(locale, dateStyle)
  const timePattern = getTimeLocalePattern(locale, timeStyle)
  const dateTimePattern = getPattern(locale, dateStyle, 'dateTime')
  return dateTimePattern.replace('{1}', datePattern).replace('{0}', timePattern)
}


lib/code.d.ts
export type PatternFormat = 'date' | 'time' | 'dateTime'

export type PatternStyle = 'full' | 'long' | 'medium' | 'short'

export interface PatternData {
  formats: PatternFormat[]
  styles: PatternStyle[]
  patterns: {
    [locale: string]: string[][]
  }
}

export interface DateOptions {
  dateStyle: PatternStyle
}

export interface TimeOptions {
  timeStyle: PatternStyle
}

export interface DateTimeOptions {
  dateStyle: PatternStyle
  timeStyle: PatternStyle
}

export function setLocalePatterns(data: PatternData): void
export function getDateLocalePattern(locale: Intl.DateTimeFormat | string,
  style?: PatternStyle | DateOptions): string
export function getTimeLocalePattern(locale: Intl.DateTimeFormat | string,
  style?: PatternStyle | TimeOptions): string
export function getDateTimeLocalePattern(locale: Intl.DateTimeFormat | string,
  dateStyle?: PatternStyle | DateTimeOptions, timeStyle?: PatternStyle): string


rollup.config.js
import json from '@rollup/plugin-json'
import cleanup from 'rollup-plugin-cleanup'
import { minify } from 'rollup-plugin-swc-minify'

export default [
  {
    input: 'lib/index.js',
    output: {
      file: 'lib/index.cjs',
      format: 'cjs',
      sourcemap: true,
    },
    external: './all.json',
    plugins: [
      cleanup()
    ]
  },
  {
    input: 'lib/index.js',
    output: [
      {
        file: 'lib/index.min.js',
        format: 'umd',
        name: 'dateTimeLocalePatterns',
        sourcemap: true,
        plugins: [
          minify()
        ]
      }
    ],
    plugins: [
      json()
    ]
  },
  {
    input: 'lib/code.js',
    output: [
      {
        file: 'lib/code.cjs',
        format: 'cjs',
        sourcemap: true,
      },
      {
        file: 'lib/code.min.js',
        format: 'umd',
        name: 'dateTimeLocalePatterns',
        sourcemap: true,
        plugins: [
          minify()
        ]
      }
    ],
    plugins: [
      cleanup()
    ]
  },
  {
    input: 'lib/short.js',
    output: {
      file: 'lib/short.min.js',
      format: 'umd',
      name: 'dateTimeLocalePatternsShort',
      sourcemap: true,
      plugins: [
        minify()
      ]
    },
    plugins: [
      json()
    ]
  }
]


perf/speed/index.js
import createSuite from './createSuite.js'
import dateAndTime from 'date-and-time'
import { format as dateFnsFormat } from 'date-fns'
import { DateTime } from 'luxon'

const { format: dateAndTimeFormat, compile: dateAndTimeCompile } = dateAndTime
const { DATETIME_SHORT, DATETIME_MED, DATETIME_FULL, DATETIME_HUGE } = DateTime

const date = new Date(1, 1, 3, 16, 5, 6) // 1901-02-03 16:05:06
const luxonDate = DateTime.fromJSDate(date)

// ---------- short

const shortPattern = 'M/d/yy, h:mm a'

function dateAndTimeShort() {
  dateAndTimeFormat(date, shortPattern)
}

const shortCompiled = dateAndTimeCompile(shortPattern)

function dateAndTimeShortCompiled() {
  dateAndTimeFormat(date, shortCompiled)
}

function dateFnsShort() {
  dateFnsFormat(date, shortPattern)
}

function luxonShort() {
  DateTime.fromJSDate(date).toLocaleString(DATETIME_SHORT)
}

function luxonShortCreated() {
  luxonDate.toLocaleString(DATETIME_SHORT)
}

// ---------- medium

const mediumPattern = 'MMM d, y, h:mm:ss a'

function dateAndTimeMedium() {
  dateAndTimeFormat(date, mediumPattern)
}

const mediumCompiled = dateAndTimeCompile(mediumPattern)

function dateAndTimeMediumCompiled() {
  dateAndTimeFormat(date, mediumCompiled)
}

function dateFnsMedium() {
  dateFnsFormat(date, mediumPattern)
}

function luxonMedium() {
  DateTime.fromJSDate(date).toLocaleString(DATETIME_MED)
}

function luxonMediumCreated() {
  luxonDate.toLocaleString(DATETIME_MED)
}

// ---------- long

const longPattern = 'MMMM d, y, h:mm:ss a z'

function dateAndTimeLong() {
  dateAndTimeFormat(date, longPattern)
}

const longCompiled = dateAndTimeCompile(longPattern)

function dateAndTimeLongCompiled() {
  dateAndTimeFormat(date, longCompiled)
}

function dateFnsLong() {
  dateFnsFormat(date, longPattern)
}

function luxonLong() {
  DateTime.fromJSDate(date).toLocaleString(DATETIME_FULL)
}

function luxonLongCreated() {
  luxonDate.toLocaleString(DATETIME_FULL)
}

// ---------- full

const fullPattern = 'EEEE, MMMM d, y, h:mm:ss a zzzz'

function dateAndTimeFull() {
  dateAndTimeFormat(date, fullPattern)
}

const fullCompiled = dateAndTimeCompile(fullPattern)

function dateAndTimeFullCompiled() {
  dateAndTimeFormat(date, fullCompiled)
}

function dateFnsFull() {
  dateFnsFormat(date, fullPattern)
}

function luxonFull() {
  DateTime.fromJSDate(date).toLocaleString(DATETIME_HUGE)
}

function luxonFullCreated() {
  luxonDate.toLocaleString(DATETIME_HUGE)
}

// ---------- suites

createSuite('Formatting a short pattern with...')
  .add('date-and-time', dateAndTimeShort)
  .add('date-and-time compiled', dateAndTimeShortCompiled)
  .add('date-fns', dateFnsShort)
  .add('luxon', luxonShort)
  .add('luxon created', luxonShortCreated)
  .run()

console.log()

createSuite('Formatting a medium pattern with...')
  .add('date-and-time', dateAndTimeMedium)
  .add('date-and-time compiled', dateAndTimeMediumCompiled)
  .add('date-fns', dateFnsMedium)
  .add('luxon', luxonMedium)
  .add('luxon created', luxonMediumCreated)
  .run()

console.log()

createSuite('Formatting a long pattern with...')
  .add('date-and-time', dateAndTimeLong)
  .add('date-and-time compiled', dateAndTimeLongCompiled)
  .add('date-fns', dateFnsLong)
  .add('luxon', luxonLong)
  .add('luxon created', luxonLongCreated)
  .run()

console.log()

createSuite('Formatting a full pattern with...')
  .add('date-and-time', dateAndTimeFull)
  .add('date-and-time compiled', dateAndTimeFullCompiled)
  .add('date-fns', dateFnsFull)
  .add('luxon', luxonFull)
  .add('luxon created', luxonFullCreated)
  .run()
